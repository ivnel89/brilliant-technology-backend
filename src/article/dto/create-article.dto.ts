import { ApiProperty } from "@nestjs/swagger";
import { IsString, IsUUID } from "class-validator";

export class CreateArticleDto {
    @ApiProperty()
    @IsString()
    content: string;
  
    @ApiProperty()
    @IsString()
    @IsUUID()
    authorId: string;
}
