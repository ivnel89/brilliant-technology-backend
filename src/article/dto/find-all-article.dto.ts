import { ApiProperty } from "@nestjs/swagger";
import { IsNumber, IsOptional, IsUUID, Max } from "class-validator";

export class FindAllArticleDto {
    @IsOptional()
    @IsNumber()
    @ApiProperty({
      required: false
    })
    offset?: number;

    @IsOptional()
    @IsNumber()
    @Max(250)
    @ApiProperty({
      required: false
    })
    limit?: number;

    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  