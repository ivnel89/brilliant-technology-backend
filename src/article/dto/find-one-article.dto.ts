import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsUUID } from "class-validator";

export class FindOneArticleDto {
    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  