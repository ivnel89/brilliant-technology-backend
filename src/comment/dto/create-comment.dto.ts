import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUUID, Length } from "class-validator";

export class CreateCommentDto {
  @ApiProperty()
  @IsString()
  @Length(1, 10000)
  content: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  authorId: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  articleId: string;

  @ApiProperty({
    nullable: true,
    required: false
  })
  @IsString()
  @IsUUID()
  @IsOptional()
  parentCommentId: string;
}
