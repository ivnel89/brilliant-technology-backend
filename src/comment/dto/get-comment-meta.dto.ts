import { ApiProperty } from "@nestjs/swagger";
import { IsArray, IsString, IsUUID } from "class-validator";

export class GetCommentsDto {
  @ApiProperty()
  @IsString()
  @IsUUID()
  userId: string;

  @ApiProperty({
     isArray: true
  })
  @IsArray()
  commentIds: Array<string>;
}
