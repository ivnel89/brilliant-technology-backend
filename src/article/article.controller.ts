import { Controller, Get, Post, Body, Patch, Param, Delete, Query } from '@nestjs/common';
import { ArticleService } from './article.service';
import { CreateArticleDto } from './dto/create-article.dto';
import { FindAllArticleDto } from './dto/find-all-article.dto';

@Controller('article')
export class ArticleController {
  constructor(private readonly articleService: ArticleService) {}

  @Post()
  create(@Body() createArticleDto: CreateArticleDto) {
    return this.articleService.create(createArticleDto);
  }

  @Get()
  findAll(
    @Query() findAllArticleDto?: FindAllArticleDto,
  ) {
    return this.articleService.findAll(findAllArticleDto);
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.articleService.findOne(id);
  }  
}
