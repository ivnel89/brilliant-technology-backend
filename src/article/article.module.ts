import { Module } from '@nestjs/common';
import { ArticleService } from './article.service';
import { ArticleController } from './article.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { UserModule } from 'src/user/user.module';
import { Article } from './entities/article.entity';

@Module({
  imports: [
    TypeOrmModule.forFeature([Article]),
    UserModule
  ],
  controllers: [ArticleController],
  providers: [ArticleService],
  exports: [ArticleService]
})
export class ArticleModule {}
