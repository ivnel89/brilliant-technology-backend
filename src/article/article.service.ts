import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { UserService } from 'src/user/user.service';
import { Repository } from 'typeorm';
import { CreateArticleDto } from './dto/create-article.dto';
import { FindAllArticleDto } from './dto/find-all-article.dto';
import { Article } from './entities/article.entity';

@Injectable()
export class ArticleService {
  constructor(
    @InjectRepository(Article)
    private articleRepository: Repository<Article>,
    private readonly userService: UserService
  ) {}

  async create(createArticleDto: CreateArticleDto) {
    const user = await this.userService.findOne(createArticleDto?.authorId);
    return this.articleRepository.save(
      new Article(user, createArticleDto?.content),
    );
  }

  findAll(findAllArticleDto: FindAllArticleDto): Promise<Array<Article>> {
    const { limit = 10, offset = 0} = findAllArticleDto
    return this.articleRepository.find({
      order: {
        lastModifiedDate: 'ASC',
      },
      skip: offset,
      take: limit,
    });  }

  findOne(id: string): Promise<Article> {
    return this.articleRepository.findOne({
      where: {
        id,
      },
    });  
  }
}
