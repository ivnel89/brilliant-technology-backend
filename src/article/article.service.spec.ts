import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { mockUser, UserRepository } from 'src/user/test/user.repository.mock';
import { UserService } from 'src/user/user.service';
import { Repository } from 'typeorm';
import { ArticleService } from './article.service';
import { Article } from './entities/article.entity';
import { createArticleDto, findAllArticleDto, mockArticleId } from './test/article.dto.mock';
import { ArticleRepository } from './test/article.repository.mock';

describe('ArticleService', () => {
  let service: ArticleService;
  let repository: Repository<Article>;
  let userService: UserService;
  let repositorySaveSpy, repositoryFindSpy, repositoryFindOneSpy;
  let userServiceFindOneSpy

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        ArticleService,
        ArticleRepository,
        UserService,
        UserRepository
      ],
    }).compile();

    service = module.get<ArticleService>(ArticleService);
    repository = module.get<Repository<Article>>(
      getRepositoryToken(Article),
    );
    userService = module.get<UserService>(UserService);
    repositorySaveSpy = jest.spyOn(repository,'save');
    repositoryFindSpy = jest.spyOn(repository,'find');
    repositoryFindOneSpy = jest.spyOn(repository,'findOne');
    userServiceFindOneSpy = jest.spyOn(userService,'findOne');

  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('create', () => {
    const saveMockError = 'saveMockError';

    beforeEach(() => {
      userServiceFindOneSpy.mockResolvedValueOnce(mockUser);
    })

    it("should call repository's save method with right params", async () => {
      await expect(service.create(createArticleDto))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(userServiceFindOneSpy).toBeCalledWith(createArticleDto.authorId)
          expect(repositorySaveSpy).toBeCalledWith(
            new Article(
              mockUser,
              createArticleDto.content
            ),
          );
        });
    });

    it('should throw an error when save throws an error', async () => {
      repositorySaveSpy.mockRejectedValueOnce(saveMockError);
      await expect(service.create(createArticleDto)).rejects.toBe(saveMockError);
    });
  });

  describe('findAll', () => {
    const findMockError = 'findMockError';

    it("should call repository's find method with right params", async () => {
      await expect(service.findAll(findAllArticleDto))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositoryFindSpy).toBeCalledWith({
            order: {
              lastModifiedDate: 'ASC',
            },
            skip: findAllArticleDto.offset,
            take: findAllArticleDto.limit,
          });
        });
    });

    it('should throw an error when find throws an error', async () => {
      repositoryFindSpy.mockRejectedValueOnce(findMockError);
      await expect(service.findAll(findAllArticleDto)).rejects.toBe(findMockError);
    });
  });

  describe('findOne', () => {
    const findOneMockError = 'findOneMockError';

    it("should call repository's findOne method with right params", async () => {
      await expect(service.findOne(mockArticleId))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositoryFindOneSpy).toBeCalledWith({
            where: {
              id: mockArticleId,
            },
          });
        });
    });

    it('should throw an error when findOne throws an error', async () => {
      repositoryFindOneSpy.mockRejectedValueOnce(findOneMockError);
      await expect(service.findOne(mockArticleId)).rejects.toBe(findOneMockError);
    });
  });
});
