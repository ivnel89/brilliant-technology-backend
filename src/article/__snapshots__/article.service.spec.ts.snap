// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleService create should call repository's save method with right params 1`] = `
Article {
  "author": User {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "isActive": true,
    "lastName": "Bar",
  },
  "content": "Foo Bar",
}
`;

exports[`ArticleService findAll should call repository's find method with right params 1`] = `
Array [
  Article {
    "author": User {
      "displayPicture": "www.example.com/img.png",
      "firstName": "Foo",
      "isActive": true,
      "lastName": "Bar",
    },
    "comments": Array [
      Comment {
        "author": User {
          "displayPicture": "www.example.com/img.png",
          "firstName": "Foo",
          "isActive": true,
          "lastName": "Bar",
        },
        "content": "comment content",
        "parentArticle": [Circular],
      },
    ],
    "content": "article content",
  },
]
`;

exports[`ArticleService findOne should call repository's findOne method with right params 1`] = `
Article {
  "author": User {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "isActive": true,
    "lastName": "Bar",
  },
  "comments": Array [
    Comment {
      "author": User {
        "displayPicture": "www.example.com/img.png",
        "firstName": "Foo",
        "isActive": true,
        "lastName": "Bar",
      },
      "content": "comment content",
      "parentArticle": [Circular],
    },
  ],
  "content": "article content",
}
`;
